GEM and ME0 segments into trackerMuon

jshlee opened the pull request:
This PR adds GemMuon and Me0Muon as well as adding GEMSegments and ME0Segments into trackerMuons.

TAMuonSegmentMatch and TrackAssociatorParameters
- added GEM segments
- added ME0 segments

MuonIdProducer 
- added isGEMMuon function
- added isME0Muon function
- save isME0Muon and isGEMMuon with trackerMuons

muon.h
- added numberOfChambersNoGEM function
- added numberOfChambersNoME0 function
- added isGEMMuon function
- added isME0Muon function

MuonChamberMatch, MuonDetIdAssociator, DetIdInfo, TAMuonChamberMatch, TrackDetectorAssociator
- added ME0 and GEM 

ME0SegmentBuilder
- fix chamberId()

MuonME0DetLayerGeometryBuilder
- bug fix, no layer 0

jshlee committed:
Merged GEMSegments-81X from repository jshlee

jshlee committed:
added gemsegment matching to trackerMuons

jshlee committed:
revert back RecoLocalMuon_cff.py

jshlee committed:
revert RecoLocalMuon_EventContent_cff.py

jshlee committed:
Merged trackerGEM-81X from repository jshlee

jshlee committed:
me0muon - added me0 segments to trackerMuon like gem

jshlee committed:
bug fix

jshlee committed:
corrected me0 muons

jshlee committed:
muon.h small correction

jshlee committed:
clean up

jshlee committed:
cleanup

jshlee committed:
removed unused functions

jshlee committed:
reverting MuonSelectors

cmsbuild commented:
A new Pull Request was created by  @jshlee (Jason Lee) for CMSSW_8_1_X.

It involves the following packages:

DataFormats/MuonReco
RecoLocalMuon/GEMSegment
RecoMuon/DetLayers
RecoMuon/MuonIdentification
TrackingTools/TrackAssociator


@cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@battibass, @makortel, @abbiendi, @GiacomoSguazzoni, @jhgoh, @VinInn, @bellan, @HuguesBrun, @rovere, @amagitte, @gpetruc, @trocino, @dgulhan, @bachtis, @rociovilar this is something you requested to watch as well.
@slava77, @degano, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

jshlee commented:
@calabria and @kpedro88 - PR for GEMMuon and ME0Muon

kpedro88 commented:
@jshlee - reminder: at some point, we need a long-term fix for MuonSegFit:
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/RecoLocalMuon/GEMSegment/plugins/MuonSegFit.cc#L310

slava77 commented:
@jshlee 
is there a presentation/slides in muon POG that describe the changes?
please add to the PR description

kpedro88 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
Can the isRPC be renamed to isGEMRPC so that we don't generate a new type?

slava77 commented:
also, please add some discussion what was already done in
https://github.com/cms-sw/cmssw/pull/13308
and is now present here and what was changed.

jshlee commented:
@slava77  - no yet for muon POG, we will give one soon, but due to the ECFA deadlines we wanted to go ahead with this first. I'm not sure what you mean by isRPC.

@kpedro88 - yes, will fix this soon.

slava77 commented:
On 6/9/16 8:06 AM, Jason Lee wrote:
> @slava77 <https://github.com/slava77> - no yet for muon POG, we will
> give one soon, but due to the ECFA deadlines we wanted to go ahead with
> this first. I'm not sure what you mean by isRPC.

RPC type technically covers and doesn't overlap in meaning with GEM
(an extra layer of measurement in the same station as CSC)

> 
> @kpedro88 <https://github.com/kpedro88> - yes, will fix this soon.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14833#issuecomment-224924569>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcburwBtElSfvG_cTSrj3O5uhwQ2wlks5qKCvigaJpZM4IyDkH>.
>

slava77 commented:
Muon POG should sign off on this code.
a message from the conveners should be enough.

cmsbuild commented:
-1

Tested at: 14f997dbe87314dc0a7f71bc65e8c60e7dc528a5

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13444/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
134.911 step2
<pre>runTheMatrix-results/134.911_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_2016+RECODR2_2016reHLT+HARVESTDR2/step2_RunSinglePh2015D+RunSinglePh2015D+HLTDR2_2016+RECODR2_2016reHLT+HARVESTDR2.log

</pre>1306.0 step2
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step2_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>135.4 step1
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step1_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>1330.0 step2
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step2_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>25202.0 step2
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-09-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Jun  9 20:34:36 2016-date Thu Jun  9 20:32:33 2016 s - exit: 18688
cmsDriver.py RelVal -s HLT:PRef,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_PRef --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_PRef_DATA.root --fileout file:RelVal_Raw_PRef_DATA_HLT_RECO.root : FAILED - time: date Thu Jun  9 20:34:36 2016-date Thu Jun  9 20:32:33 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-09-1100/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Jun  9 20:46:08 2016-date Thu Jun  9 20:32:36 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-09-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Jun  9 20:50:44 2016-date Thu Jun  9 20:32:38 2016 s - exit: 18688
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_GRun_MC.root --fileout file:RelVal_Raw_GRun_MC_HLT_RECO.root : FAILED - time: date Thu Jun  9 20:50:44 2016-date Thu Jun  9 20:32:38 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-09-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Jun  9 20:50:44 2016-date Thu Jun  9 20:32:44 2016 s - exit: 18688
cmsDriver.py RelVal -s HLT:PRef,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_PRef --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_PRef_MC.root --fileout file:RelVal_Raw_PRef_MC_HLT_RECO.root : FAILED - time: date Thu Jun  9 20:50:44 2016-date Thu Jun  9 20:32:44 2016 s - exit: 18688
cmsDriver.py TTbar_13TeV_TuneCUETP8M1_cfi --conditions auto:run2_mc --fast  -n 100 --eventcontent AODSIM,DQM --relval 100000,1000 -s GEN,SIM,RECOBEFMIX,DIGI:pdigi_valid,L1,DIGI2RAW,L1Reco,RECO,EI,HLT:@relval2016,VALIDATION  --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --datatier GEN-SIM-DIGI-RECO,DQMIO --beamspot NominalCollision2015 --era Run2_2016 --magField 38T_PostLS1 : FAILED - time: date Thu Jun  9 20:41:34 2016-date Thu Jun  9 20:32:45 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-09-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Jun  9 20:35:39 2016-date Thu Jun  9 20:33:39 2016 s - exit: 18688
cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_PIon_DATA.root --fileout file:RelVal_Raw_PIon_DATA_HLT_RECO.root : FAILED - time: date Thu Jun  9 20:35:39 2016-date Thu Jun  9 20:33:39 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-09-1100/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Jun  9 20:55:05 2016-date Thu Jun  9 20:35:47 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-09-1100/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Thu Jun  9 20:48:47 2016-date Thu Jun  9 20:46:09 2016 s - exit: 18688
cmsDriver.py RelVal -s HLT:GRun,RAW2DIGI,L1Reco,RECO --data --scenario=pp -n 10 --conditions auto:run2_data_GRun --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_25ns --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_GRun_DATA.root --fileout file:RelVal_Raw_GRun_DATA_HLT_RECO.root : FAILED - time: date Thu Jun  9 20:48:47 2016-date Thu Jun  9 20:46:09 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week0/slc6_amd64_gcc530/cms/cmssw-patch/CMSSW_8_1_X_2016-06-09-1100/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Thu Jun  9 21:04:03 2016-date Thu Jun  9 20:48:50 2016 s - exit: 18688
cmsDriver.py RelVal -s HLT:PIon,RAW2DIGI,L1Reco,RECO --mc --scenario=pp -n 10 --conditions auto:run2_mc_PIon --relval 9000,50 --datatier "RAW-HLT-RECO" --eventcontent FEVTDEBUGHLT --customise=HLTrigger/Configuration/CustomConfigs.L1THLT --era Run2_2016 --magField 38T_PostLS1 --processName=HLTRECO --filein file:RelVal_Raw_PIon_MC.root --fileout file:RelVal_Raw_PIon_MC_HLT_RECO.root : FAILED - time: date Thu Jun  9 21:04:03 2016-date Thu Jun  9 20:48:50 2016 s - exit: 18688

kpedro88 commented:
@jshlee: you may need to make some of these parameters/tags part of an Era
```
----- Begin Fatal Exception 09-Jun-2016 19:50:35 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=MuonIdProducer label='hltMuons'
Exception Message:
MissingParameter: Parameter 'GEMSegmentCollectionLabel' not found.
----- End Fatal Exception -------------------------------------------------
```

slava77 commented:
@jshlee @kpedro88 
how do we test muon reco in phase2?
(IIRC the matrix workflows are at best up to RECO:tracking_only)
-w upgrade has things like 10400.0 , with pure customs configs and full RECO, but they are dead.

kpedro88 commented:
@slava77 - I don't have any central workflow for muon reco at the moment...

slava77 commented:
On 6/10/16 9:23 AM, Kevin Pedro wrote:
> @slava77 <https://github.com/slava77> - I don't have any central
> workflow for muon reco at the moment...

Is there a non-central workflow that works?

There was a premise of ECFA for this to go ahead sooner than
presentation/discussion in the POG.
but if the code can't run centrally, what's the point?

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14833#issuecomment-225229207>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbmInOsO-25nSD7AUpKkbVdJliZPWks5qKY98gaJpZM4IyDkH>.
>

kpedro88 commented:
I've been trying to work on a "full reco" workflow (using Run2 calorimeters) this week. Unfortunately, changes in the Phase2 tracking objects are causing a lot of difficulties. But the muon systems haven't given me any trouble during my tests, at least.

I'm not sure what else can be done right now. If @jshlee and @calabria have private testing configurations, hopefully they can post them.

slava77 commented:
On 6/10/16 9:27 AM, Kevin Pedro wrote:
> I've been trying to work on a "full reco" workflow (using Run2
> calorimeters) this week. Unfortunately, changes in the Phase2 tracking
> objects are causing a lot of difficulties. But the muon systems haven't
> given me any trouble during my tests, at least.
> 
> I'm not sure what else can be done right now. If @jshlee
> <https://github.com/jshlee> and @calabria <https://github.com/calabria>
> have private testing configurations, hopefully they can post them.

anything cmsDriver-able would do, since this can be used in sample
production requests.


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14833#issuecomment-225230124>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbiiQ1eWJDA-dChhj7UCra5lmSiMtks5qKZBogaJpZM4IyDkH>.
>

jshlee commented:
@slava77 and @kpedro88 
I've been testing muon reco with the cmsDriver command below
cmsDriver.py SingleMuPt100_cfi -s GEN,SIM,DIGI,L1,DIGI2RAW,RAW2DIGI,L1Reco,RECO --conditions auto:run2_mc --magField 38T_PostLS1 --datatier GEN-SIM --geometry Extended2015MuonGEMDev,Extended2015MuonGEMDevReco --eventcontent FEVTDEBUGHLT --era phase2_muon,run3_GEM -n 100 --fileout out_reco.root 

For the MissingParameter in hltMuons, should I use fillDescriptions or just update the hlt python files?

slava77 commented:
On 6/11/16 5:49 AM, Jason Lee wrote:
> @slava77 <https://github.com/slava77> and @kpedro88
> <https://github.com/kpedro88>
> I've been testing muon reco with the cmsDriver command below
> cmsDriver.py SingleMuPt100_cfi -s
> GEN,SIM,DIGI,L1,DIGI2RAW,RAW2DIGI,L1Reco,RECO --conditions auto:run2_mc
> --magField 38T_PostLS1 --datatier GEN-SIM --geometry
> Extended2015MuonGEMDev,Extended2015MuonGEMDevReco --eventcontent
> FEVTDEBUGHLT --era phase2_muon,run3_GEM -n 100 --fileout out_reco.root
> 
> For the MissingParameter in hltMuons, should I use fillDescriptions or
> just update the hlt python files?

fillDescriptions is preferred

more complicated situations are dealt with in
HLTrigger/Configuration/python/customizeHLTforCMSSW.py

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14833#issuecomment-225359364>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbmI-pSvAfRNWhOCnP79YVXWJP6MGks5qKq7UgaJpZM4IyDkH>.
>

jshlee commented:
@kpedro88 - I tried to use fillDescriptions, but it doesn't seem to get pick up properly. I need to update the TrackAssociatorParameters 3 times.
Also, when this is run on hlt, all the inputs are from the hlt collections. So for the gem and me0 segments, we should make hltGEMSegments and hltME0Segments.

kpedro88 commented:
@jshlee it's hard for me to guess what the problem might be with so little information...

If the HLT collections are necessary at this point, then yes, we should make them. But for the moment, at least we can see a "missing collection" error instead of a "missing parameter" error.

jshlee commented:
@kpedro88 - currently the problem is that the TrackAssociatorParameters should be derived from the TrackingTools.TrackAssociator.default_cfi, where GEMSegmentCollectionLabel are defined.

But, HLT does not import TrackAssociatorParameters from TrackingTools.TrackAssociator.default_cfi. 
The HLT defines everything in one python config file like OnLine_HLT_GRun.py.

Eventually, we would like to run this at the HLT as well.

kpedro88 commented:
@jshlee - I still don't fully understand the problem, but it sounds like you might have to make use of:
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/HLTrigger/Configuration/python/customizeHLTforCMSSW.py
maybe @slava77 can offer some additional advice.

slava77 commented:
I can't add more than what I mentioned in 
https://github.com/cms-sw/cmssw/pull/14833#issuecomment-225361963

jshlee commented:
@slava77 and @kpedro88 - I tried the following
` edm::ParameterSetDescription desc;  
  edm::ParameterSetDescription descNested;
  descNested.add<edm::InputTag>("GEMSegmentCollectionLabel",edm::InputTag("gemSegments"));
  descNested.add<edm::InputTag>("ME0SegmentCollectionLabel",edm::InputTag("me0Segments"));
  descNested.setAllowAnything();
  desc.add<edm::ParameterSetDescription>("TrackAssociatorParameters", descNested);  
  desc.setAllowAnything();
  descriptions.addDefault(desc); `
but it doesn't seem to pick GEMSegmentCollectionLabel up.
any ideas on how I can fix this?

kpedro88 commented:
@jshlee - that looks correct to me... I'll try to think about why it might be failing, but I'm not an expert on how the HLT configs get generated.

slava77 commented:
@jshlee 
``` but it doesn't seem to pick GEMSegmentCollectionLabel up. ```
what did you mean by that?

jshlee commented:
@slava77 - it still complains that GEMSegmentCollectionLabel is missing from hltMuons

jshlee commented:
@slava77 and @kpedro88 - I solved the issue... there were more nested copies of `TrackAssociatorParameters`.

jshlee committed:
added fillDescriptions for MuonIdProducer and MuonDetIdAssociator

cmsbuild commented:
Pull request #14833 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
@kpedro88 - I've updated this PR with the fillDescriptions. Can you test please.

kpedro88 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: a4a6342e6a190d6ee635369a749aeaa84d652c8e

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13476/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals  AddOn**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
1330.0 step2
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15/step2_ZMM_13+ZMM_13+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>25202.0 step2
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25/step2_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25.log

</pre>

* **AddOn**:

I found errors in the following addon tests:

cmsRun /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-06-12-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Tue Jun 14 00:04:01 2016-date Mon Jun 13 23:55:54 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-06-12-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Tue Jun 14 00:02:46 2016-date Mon Jun 13 23:55:58 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-06-12-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Tue Jun 14 00:05:15 2016-date Mon Jun 13 23:55:59 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-06-12-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PRef.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Tue Jun 14 00:04:24 2016-date Mon Jun 13 23:56:07 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-06-12-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Tue Jun 14 00:00:44 2016-date Mon Jun 13 23:56:13 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-06-12-2300/src/HLTrigger/Configuration/test/OnLine_HLT_HIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Tue Jun 14 00:04:14 2016-date Mon Jun 13 23:56:20 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-06-12-2300/src/HLTrigger/Configuration/test/OnLine_HLT_GRun.py realData=True  globalTag=@ inputFiles=@  : FAILED - time: date Tue Jun 14 00:02:31 2016-date Mon Jun 13 23:56:34 2016 s - exit: 18688
cmsRun /cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc530/cms/cmssw/CMSSW_8_1_X_2016-06-12-2300/src/HLTrigger/Configuration/test/OnLine_HLT_PIon.py realData=False globalTag=@ inputFiles=@  : FAILED - time: date Tue Jun 14 00:04:32 2016-date Mon Jun 13 23:56:42 2016 s - exit: 18688

makortel commented:
(sorry to jump in without knowing too much of the details)

The last error appears to be caused by the fact that `fillDescriptions()` were added to ESProduct (`MuonDetIdAssociator`), not to ESProducer (`DetIdAssociatorESProducer`). Unfortunately the `DetIdAssociatorESProducer` uses the "helper plugins via factory" pattern, for which there is currently no easy/straightforward solution for `fillDescriptions()` (AFAIK; we have plenty of those in tracking code). At least for this case you're probably better off with https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/HLTrigger/Configuration/python/customizeHLTforCMSSW.py.

jshlee commented:
thanks @makortel! I'll think of another way to fix this.

cmsbuild commented:
Pull request #14833 was updated. @cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
Since `fillDescriptions()` isn't an option, I used `try` to find gem and me0 detectors. 
Let me know if `customizeHLTforCMSSW.py` is preferred.

slava77 commented:
try-catch is a bad solution in this context.

jshlee committed:
customizeHLTforCMSSW.py added customiseForMuon
to add in includeGEM and includeME0 parameters

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @Martin-Grunewald, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
@slava77 - I changed to `customizeHLTforCMSSW.py`

fwyzard commented:
-1

fwyzard commented:
- entries in `customizeHLTforCMSSW.py` should be named after a specific pull request number, for ease of identification
  - the calls to the customisation functions should be dependent on the CMSSW version, to avoid problems when we backport the same file from 8.1.x to 8.0.x

fwyzard commented:
A note for the framework export: if `fillDescriptions()` is currently not an option for modules use the "helper plugins via factory" pattern, the plugin mechanism should be extended to support it.

jshlee committed:
updated customizeHLTforCMSSW with Andrea Bocci comments

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @Martin-Grunewald, @slava77, @davidlange6 can you please check and sign again.

slava77 commented:
there are merge conflicts here
please fix

jshlee committed:
Merge branch 'merge-attempt' into jshlee/trackerGEM-81X

Conflicts:
	HLTrigger/Configuration/python/customizeHLTforCMSSW.py

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @Martin-Grunewald, @slava77, @davidlange6 can you please check and sign again.

jshlee commented:
fixed conflicts

jshlee committed:
Merged trackerGEM-81X from repository jshlee

slava77 commented:
@jshlee please confirm that ```runTheMatrix.py -l limited ``` runs

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
@battibass, @HuguesBrun
Please confirm that these code changes were discussed within muon POG and that they are good to go from the muon POG point of view.
As of the PR submission, this wasn't yet presented in a meeting.

slava77 commented on DataFormats/MuonReco/interface/Muon.h:
```diff
@@ -179,15 +179,20 @@ namespace reco {
+    /// number of chambers not including GEM matches (MuonChamberMatches include RPC rolls, and also GEM)
+    int numberOfChambersNoGEM() const;    
+    /// number of chambers not including ME0 matches (MuonChamberMatches include RPC rolls, and also ME0)
+    int numberOfChambersNoME0() const;    
```
the purpose of numberOfChambersNoRPC was to capture the number of CSC-or-DT chambers.
I picked the name in 2011 not thinking of the future upgrades for a somewhat obvious reason.

numberOfChambersNoGEM and numberOfChambersNoME0 don't fit this purpose.
To make a more meaningful update, rename numberOfChambersNoRPC to be numberOfChambersCSCorDT.

slava77 commented on DataFormats/MuonReco/src/Muon.cc:
```diff
@@ -74,6 +100,14 @@ int Muon::numberOfMatches( ArbitrationType type ) const
+	continue;
+      }
+      if(type == GEMSegmentAndTrackArbitration) {
+	if (chamberMatch->id.subdetId() == MuonSubdetId::GEM) matches += chamberMatch->segmentMatches.size();
```
this is a bogus arbitration.
Do you really want to keep it this way?

jshlee replied:
This is just a placeholder until the me0 and gem performance has been studied. For now we just need number of ME0 or GEM matches.

slava77 commented on RecoMuon/DetLayers/src/MuonME0DetLayerGeometryBuilder.cc:
```diff
@@ -32,7 +32,7 @@ MuonME0DetLayerGeometryBuilder::buildEndcapLayers(const ME0Geometry& geo) {
     int iendcap = (endcap==1) ? 0 : 1; // +1: forward, -1: backward
 
-    for(int layer = ME0DetId::minLayerId; layer <= ME0DetId::maxLayerId; ++layer) { 
+    for(int layer = ME0DetId::minLayerId+1; layer <= ME0DetId::maxLayerId; ++layer) { 
```
please clarify what this fixes

jshlee replied:
ME0 has 6 layers, minLayerId = 0 and maxLayerId = 6, so it was making one extra layer.

slava77 commented on RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc:
```diff
@@ -752,6 +761,18 @@ bool MuonIdProducer::isGoodRPCMuon( const reco::Muon& muon )
+
+bool MuonIdProducer::isME0Muon( const reco::Muon& muon )
+{
+  if(muon.track()->pt() < minPt_ || muon.track()->p() < minP_) return false;
```
do you really want to apply the same pt cut for ME0 as used for CSC?
This value was selected/tuned to accept all muons within CSC acceptance.
ME0 goes beyond that.
pt of 0.5 GeV is probably low enough, for eta=3 this corresponds to p of 5 GeV. 
This is probably about 2 GeV above the rangeout.

jshlee replied:
Yep, your right. I think the pt cut needs to be studied for both.
Shall I remove them for now and add in appropriate ones later?

slava77 replied:
minP of 3 GeV is probably going to be OK here.
Just removing the selection will bump up the CPU time.
An "easy" option is to leave it as is, but add a comment that this should be reviewed once we get to physics studies.

jshlee replied:
the values of minPt = cms.double(0.5) and minP = cms.double(2.5).
should I just use the minP for me0?

slava77 replied:
sounds reasonable.

slava77 commented on TrackingTools/TrackAssociator/python/DetIdAssociatorESProducer_cff.py:
```diff
@@ -36,12 +36,18 @@
 eras.run2_common.toModify( muonDetIdAssociator, includeBadChambers = True )
 
+# include GEM & ME0 for phase2
+eras.phase2_muon.toModify( muonDetIdAssociator, includeGEM = True )
```
this should be also in run3_GEM (if you add to run3_GEM, the change in phase2_muon is not necessary)

slava77 commented on TrackingTools/TrackAssociator/src/TrackAssociatorParameters.cc:
```diff
@@ -63,6 +65,8 @@ void TrackAssociatorParameters::loadParameters( const edm::ParameterSet& iConfig
      dtSegmentsToken=iC.consumes<DTRecSegment4DCollection>(theDTRecSegment4DCollectionLabel);
      cscSegmentsToken=iC.consumes<CSCSegmentCollection>(theCSCSegmentCollectionLabel);
+     gemSegmentsToken=iC.consumes<GEMSegmentCollection>(theGEMSegmentCollectionLabel);
+     me0SegmentsToken=iC.consumes<ME0SegmentCollection>(theME0SegmentCollectionLabel);
```
please add these to consumes only when these detectors are enabled

slava77 commented on TrackingTools/TrackAssociator/src/TrackDetectorAssociator.cc:
```diff
@@ -750,6 +755,11 @@ void TrackDetectorAssociator::fillMuon( const edm::Event& iEvent,
+   edm::Handle<GEMSegmentCollection> gemSegments;
+   iEvent.getByToken(parameters.gemSegmentsToken, gemSegments );
+   edm::Handle<ME0SegmentCollection> me0Segments;
+   iEvent.getByToken(parameters.me0SegmentsToken, me0Segments );
```
better call these only if these detectors logic is enabled in the configuration.

cmsbuild commented:
+1
Tested at: d32c122edf653f6de2a76a7da85c38d0264b4fdb
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13507/summary.html

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/d32c122edf653f6de2a76a7da85c38d0264b4fdb
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13507/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13507/git-merge-result

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13507/summary.html

jshlee committed:
numberOfChambersNoRPC -> numberOfChambersCSCorDT
rm numberOfChambersNoGEM and numberOfChambersNoME0

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @dmitrijus, @Martin-Grunewald, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

jshlee committed:
MuonSegmentMatch added gem and me0 segments
TrackAssociatorParameters added useGEM and useME0
TrackDetectorAssociator use useGEM and useME0 to get gem and me0
segments
DetIdAssociatorESProducer_cff.py change gem to run3
muons1stStep_cfi.py added eras for run3 gem and phase2 me0

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @dmitrijus, @Martin-Grunewald, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

jshlee committed:
reco::MuonSegmentMatch updated ClassVersion

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @dmitrijus, @Martin-Grunewald, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

kpedro88 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

jshlee committed:
MuonIdProducer.cc add min pt on GEM and ME0

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @dmitrijus, @Martin-Grunewald, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

jshlee committed:
MuonIdProducer::isME0Muon change p cut to minP_

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @dmitrijus, @Martin-Grunewald, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 323755bc7c820324e55018781a2082414f63ca8b
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13529/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13529/summary.html

slava77 commented on DataFormats/MuonReco/src/classes_def.xml:
```diff
@@ -47,8 +47,8 @@
-  <class name="reco::MuonSegmentMatch" ClassVersion="10">
-   <version ClassVersion="10" checksum="754193003"/>
+  <class name="reco::MuonSegmentMatch" ClassVersion="11">
+   <version ClassVersion="11" checksum="2861296960"/>
```
old version-checksum line for version 10 should stay

slava77 commented on RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc:
```diff
@@ -752,6 +761,20 @@ bool MuonIdProducer::isGoodRPCMuon( const reco::Muon& muon )
 
+bool MuonIdProducer::isGEMMuon( const reco::Muon& muon )
+{
+  // need to update min cuts on pt
```
GEM is in the same acceptance as CSC.
So, the same p/pt cuts should still  work.

jshlee replied:
i use both pt and p for GEM

slava77 replied:
My comment was about the comment line ```// need to update min cuts on pt ```

jshlee committed:
update DataFormats/MuonReco/src/classes_def.xml

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @dmitrijus, @Martin-Grunewald, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

jshlee committed:
MuonIdProducer::isGEMMuon
comment on comment

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @dmitrijus, @Martin-Grunewald, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 2b0b062fecc335eae11a0292863d572af1fc8ea0

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13560/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test testRecoMETMETProducers had ERRORS

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13560/summary.html

jshlee commented:
@slava77 and @kpedro88, can you retest please. The errors reported were from disk read failure.

kpedro88 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 2b0b062fecc335eae11a0292863d572af1fc8ea0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13595/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13595/summary.html

slava77 commented:
@jshlee @battibass @calabria @HuguesBrun 
what do we expect from this code: should the ME0 and GEM segments be counted in the definition of isTrackerMuon?
or was the intent to simply add the GEM and ME0 segment match information without modifying the isTrackerMuon?

e.g. in FourMuFlat wf 11200.0 (I'm running on top of #14865 )
![all_sign728vssign727_fourmuflat2023wf11200p0c_recomuons_muons__reco_obj_eta2](https://cloud.githubusercontent.com/assets/4676718/16214851/5a69a962-3711-11e6-8c33-d82628ca5a24.png)
we are now getting extra tracker muons
The number of PF muons goes up as well
![all_sign728vssign727_fourmuflat2023wf11200p0c_recopfcandidates_particleflow__reco_obj_eta32](https://cloud.githubusercontent.com/assets/4676718/16214885/a9538322-3711-11e6-9fc0-c4f339ab3f0c.png)

If I'm not mistaken, the way these muons make it to PF is by having numberOfMatches >= 2 (all other selections used in PFMuonAlgo based on TM\* selectors should be blind to GEM or ME0). In the setup on top of the current #14865  the ME0 is probably broken at the geometry level, so, I think that the increased counts of muons are from GE2/1 (more likely than from GE1/1).

jshlee commented:
@slava @calabria - I understood that one of the goals is to have an Additional layer of “matched muons” using all the available segments in the muon system (DT, CSC, GEM) https://twiki.cern.ch/twiki/bin/viewauth/CMS/CMSMuonUpgradeSimulation. 
The way it is in this PR, the trackerMuon becomes this. Or else, I can add in a trackerMuon+ and keep trackerMuon unchanged.

calabria commented:
@jshlee the idea is to have GEMMuon, ME0Muon, and also a new layer of matched muons where we use all the available segments. But as we agreed with @battibass @HuguesBrun and @trocino it is better to leave the tracker muon unchanged and another trackerMuon+

slava77 commented:
The easiest here would be to move the phase2 detectors to phase2Matches_ or something with a similar name vector of MuonChamberMatches. In this case the current selection methods will not pick them up.
The alternative is to go through all of the Muon and MuonSelectors code and make sure that functions apply only to DT and CSC detectors by default and can do something else.
From the DataFormats evolution point of view, this latter solution might be better though, but the selectors code is not the easiest to edit.

trocino commented:
Yes, I agree, the trackerMuons should stay unchanged. So "isGoodTrackerMuon" should use the newly introduced "numberOfChambersCSCorDT" function, I guess. 

In theory, this problem should affect only the non-arbitrated trackerMuons, because the arbitration algorithms check only CSC and DT segments (apart from specific RPC or GEM arbitrations). So, in principle, all the practical use-cases should already be safe in the current version. 
But then I don't understand why the PFMuons are affected, I thought that only arbitrated trackerMuons could become PFMuons. Can you make sure that arbitrated trackerMuons are unchanged with this PR, @jshlee ?

slava77 commented:
On 6/21/16 7:47 AM, Daniele Trocino wrote:
> Yes, I agree, the trackerMuons should stay unchanged. So
> "isGoodTrackerMuon" should use the newly introduced
> "numberOfChambersCSCorDT" function, I guess.
> 
> In theory, this problem should affect only the non-arbitrated
> trackerMuons, because the arbitration algorithms check only CSC and DT
> segments (apart from specific RPC or GEM arbitrations). So, in
> principle, all the practical use-cases should already be safe in the
> current version.
> But then I don't understand why the PFMuons are affected, I thought that
> only arbitrated trackerMuons could become PFMuons. Can you make sure
> that arbitrated trackerMuons are unchanged with this PR, @jshlee
> <https://github.com/jshlee> ?


There are no checks for DT or CSC when arbitrated matches are counted
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_0_pre7/DataFormats/MuonReco/src/Muon.cc#L86
Segment masks are be filled irrespective of the detector
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_0_pre7/RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc#L941
well, there are some checks, but they don't capture all cases


> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14833#issuecomment-227463446>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbmV2EAHDkzmcXPguWZfO0pst2_s6ks5qN_mKgaJpZM4IyDkH>.
>

trocino commented:
You are right Slava, I got confused. It's the muon selector that uses only CSC and DT segments (because of the hard-coded range of this loop [1]), not the arbitration. 
In particular, "numberOfMatches" is affected by GEM and ME0 segments, but "numberOfMatchedStations" is not, for the same reason [2]. 

Concerning the main use-cases, this is the summary (if I'm not not much mistaken): 
- Tight, Soft and HighPt muons are not affected by GEM and ME0, because they either use "numberOfMatchedStations" or "isGoodMuon"; 
- Loose and Medium muons, and in general all PFMuons without further ID, are affected by the presence of GEM and ME0 matches. 

Would it be easier to make new arbitration types (in "numberOfMatches") that use only CSC and DT? Like "NoArbitrationCSCandDTonly" or so. This one in part. could be used inside MuonIdProducer to flag TrackerMuons. 
This change would not be transparent though. Apart from muon identification, "numberOfMatches" may be used also in other parts of the code, like validation and DQM, or even in private code.  

[1] https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/DataFormats/MuonReco/src/MuonSelectors.cc#L338 
[2] https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_0_pre7/DataFormats/MuonReco/src/Muon.cc#L116

jshlee commented:
@slava77 and @trocino - How about 
`numberOfMatches( ArbitrationType type, bool isUpgrade=false)` and `numberOfMatchedStations( ArbitrationType type = SegmentAndTrackArbitration, bool isUpgrade=false)` and for the non isUpgrade case, filter out gem and me0 hits.

slava77 commented:
On 6/21/16 9:27 AM, Jason Lee wrote:
> @slava77 <https://github.com/slava77> and @trocino
> <https://github.com/trocino> - How about
> |numberOfMatches( ArbitrationType type, bool isUpgrade=false)| and
> |numberOfMatchedStations( ArbitrationType type =
> SegmentAndTrackArbitration, bool isUpgrade=false)| and for the non
> isUpgrade case, filter out gem and me0 hits.
> 

Some mask to enable subdetectors may be more efficient.

I think that eventually all methods dealing with matches and arbitration
should become aware of the upgrade.
Long term this is easier with this default parameter
instead of introducing new default interface methods with CSCorDT and such.
Some refactoring may be necessary to not have the implementation blow up
in code line count.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/14833#issuecomment-227494996>, or
> mute the thread
> <https://github.com/notifications/unsubscribe/AEdcbu9RWZF6buXnjoIjTUhckAlOBvWvks5qOBEHgaJpZM4IyDkH>.
>

Martin-Grunewald commented:
+1

slava77 commented:
@jshlee 
please update on the progress here.
From the last messages exchanged, there should be changes made in this PR.

jshlee commented:
@slava77 - we had a talk with the POG about this and are waiting for feedback from the muon reco contacts.

jshlee committed:
Merged trackerGEM-81X from repository jshlee

jshlee committed:
changed gem chamberMatch->segmentMatches to chamberMatch->gemMatches

cmsbuild commented:
Pull request #14833 was updated. @ghellwig, @covarell, @lveldere, @ianna, @cerminar, @fabozzi, @Martin-Grunewald, @rekovic, @vanbesien, @govoni, @perrotta, @civanch, @franzoni, @cmsbuild, @fwyzard, @davidlange6, @smuzaffar, @Dr15Jones, @cvuosalo, @mdhildreth, @slava77, @ggovi, @mmusich, @mulhearn, @ssekmen, @perrozzi, @thuer, @dmitrijus, @alja, @srimanob, @hengne can you please check and sign again.

jshlee committed:
Merged CMSSW_8_1_X from repository cms-sw

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @dmitrijus, @Martin-Grunewald, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

jshlee commented:
@slava77 - The conclusion was to separate the identification.
Since all the arbitration depends on MuonChamberMatch.segmentMatches, gem and me0 are no longer saved here as before and saved instead in new MuonChamberMatch.gemMatches and MuonChamberMatch.me0Matches.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 108b4a52b832445e15f20e49121668d0da1f12ee


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/108b4a52b832445e15f20e49121668d0da1f12ee
https://github.com/cms-sw/cmssw/commit/6d4afc3a4d8400d0e088bfdf46cbeab2e62df56a
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13917/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13917/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13917/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following worklfows:
10624.0 step3
<pre>runTheMatrix-results/10624.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2023tilted_GenSimFull+DigiFull_2023tilted+RecoFullGlobal_2023tilted+HARVESTFullGlobal_2023tilted/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2023tilted_GenSimFull+DigiFull_2023tilted+RecoFullGlobal_2023tilted+HARVESTFullGlobal_2023tilted.log

</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
https://github.com/cms-sw/cmssw/commit/108b4a52b832445e15f20e49121668d0da1f12ee
https://github.com/cms-sw/cmssw/commit/6d4afc3a4d8400d0e088bfdf46cbeab2e62df56a
You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13917/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13917/git-merge-result

slava77 commented:
For a change, the error is not from DAS
workflow 10624 in step3
```
   [1] Running path 'reconstruction_step'
   [2] Calling event method for module MuonProducer/'muons'
   [3] Calling produce method for unscheduled module CosmicsMuonIdProducer/'cosmicsVeto'
Exception Message:
BadRefCore RefCore: Request to resolve a null or invalid reference to a product of type 'std::vector<reco::Track>' has been detected.
```
this is very likely due to this PR.

jshlee committed:
bugfix broken muon track ref

cmsbuild commented:
Pull request #14833 was updated. @perrotta, @cmsbuild, @cvuosalo, @fwyzard, @dmitrijus, @Martin-Grunewald, @slava77, @vanbesien, @davidlange6 can you please check and sign again.

jshlee commented:
@slava77 - I've fixed this and tested it with workflow 10624. Please test again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 46909432359f53ae88aea2305010bc72681bb80f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13935/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-14833/13935/summary.html

Martin-Grunewald commented:
+1

dmitrijus commented:
+1

slava77 commented on DataFormats/MuonReco/interface/MuonSegmentMatch.h:
```diff
@@ -46,7 +48,8 @@ namespace reco {
          CSCSegmentRef      cscSegmentRef;
-
+	 GEMSegmentRef      gemSegmentRef;
+	 ME0SegmentRef      me0SegmentRef;
```
this, together with the use case pattern is becoming more and more bloated.
``` std::vector<reco::MuonSegmentMatch> segmentMatches ``` can have a value only among dtSegmentRef or cscSegmentRef, while 
```std::vector<reco::MuonSegmentMatch> gemMatches``` can have a meaningful value only for gemSegmentRef
... and inside the MuonChamberMatch only one of these vectors has a meaning.

This is probably still OK considering the most frequent access patterns: actual segments are used mainly in detailed detector analysis (truth matching and alike).
The alternatives will likely lead to polymorphic generalized more effective storage, but, very likely, access will require casting back to specific classes.

slava77 commented:
@jshlee @calabria @trocino 
after the last updates we now have more muons in the muon collections. As expected, they are not modifying the trackerMuons set. In the (no pileup) QCD dijet sample pt 600to800 (wf 10626.0) the total number of muons about doubles (from 15 to 29).

The additional muons contribute also to the earlyMuons collection and as a result affect tracking
e.g. 10626.0
![all_sign728vsorig-not2ch_qcd600to800in13tev2023wf10626p0c_recotracks_generaltracks__reco_obj_algo](https://cloud.githubusercontent.com/assets/4676718/16792315/53003110-487c-11e6-8f45-cc7a9bcf8511.png)
Since this is only in in-out iteration, the effect is going to be rather small, but will still likely enhance low quality track reconstruction (fakes and not).
For this workfow, the effect is mainly in the GEM direction (I didn't notice changes in ME0 direction).

If you want to decouple the new tracker muon types more completely, the earlyMuons should be modified to not use GEM and ME0.

I don't think it matters much for performance, but I wanted to get a clarification what was the intended behavior (fully decoupled, just new muons, or integrated already in the muon-driven tracking iterations).

jshlee commented:
@slava77 - The idea was to go fully integrated already in the muon-driven tracking iterations, but not for identification.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/14833 	4690943
- code changes are in line with the description and the follow up review/discussion. Only phase2 (and not yet existing run3) workflows are affected.
- jenkins tests pass and show changes only in 2023 workflow
- local tests with more 2023 workflows show expected behavior: new muons appear with GEM or RPC types and contribute to inside-out tracking iteration, but not to anything else from the ```_muons_``` collection.  On-disk increase is more significant in jetty events, expected for the loose matching definitions.  On-disk increase of the reco::Muon size is negligible for non-phase-2 workflows (only empty vectors data in the new data members). CPU increase in mostly-real-muon samples is small; expected to be more significant with pileup or in high-pt jets (I'm guessing it to be less than a per cent in high-PU, even without HGCAL).

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_8_1_X IBs (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @slava77, @davidlange6, @degano, @smuzaffar

kpedro88 commented:
@davidlange6 please merge for pre9 if possible
